ipam/provider/aws.go

```diff
@@ -0,0 +1,258 @@
+	hostClusterName     string
+}
+
+type ChartValuesConfig struct {
+	AWSAPIHostedZoneGuest     string
+	AWSIngressHostedZoneGuest string
+	AWSRegion                 string
+	AWSRouteTable0            string
+	AWSRouteTable1            string
+	AWSVPCPeerID              string
+	ClusterName               string
+	CommonDomain              string
+	SSHPublicKey              string
+	VersionBundleVersion      string
+}
+
+func NewAWS(config AWSConfig) (*AWS, error) {
+	if config.AWSClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AWSClient must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AWS{
+		awsClient:     config.AWSClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+
+		chartValuesConfig: ChartValuesConfig{
+			AWSAPIHostedZoneGuest:     getValue(config.AWSAPIHostedZoneGuest, "AWS_API_HOSTED_ZONE_GUEST"),
+			AWSIngressHostedZoneGuest: getValue(config.AWSIngressHostedZoneGuest, "AWS_INGRESS_HOSTED_ZONE_GUEST"),
+			AWSRegion:                 getValue(config.AWSRegion, "AWS_REGION"),
+			AWSRouteTable0:            getValue(config.AWSRouteTable0, "AWS_ROUTE_TABLE_0"),
+			AWSRouteTable1:            getValue(config.AWSRouteTable1, "AWS_ROUTE_TABLE_1"),
+			CommonDomain:              getValue(config.CommonDomain, "COMMON_DOMAIN"),
+			SSHPublicKey:              getValue(config.SSHPublicKey, "IDRSA_PUB"),
+			VersionBundleVersion:      getValue(config.VersionBundleVersion, "VERSION_BUNDLE_VERSION"),
+		},
+		chartValuesTemplate: template.Must(template.New("awsConfigChartValues").Parse(apiextensionsAWSConfigE2EChartValues)),
+		hostClusterName:     getValue(config.ClusterName, "CLUSTER_NAME"),
+	}
+
+	return a, nil
+}
+
+func (aws *AWS) CreateCluster(clusterName string) error {
+	if clusterName == "" {
+		return microerror.Maskf(invalidConfigError, "clusterName must not be empty")
+	}
+
+	hostAWSConfig, err := aws.hostFramework.AWSCluster(aws.hostClusterName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = aws.installCertResources(clusterName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deploymentName := awsConfigDeploymentName(clusterName)
+
+	o := func() error {
+		// NOTE we ignore errors here because we cannot get really useful error
+		// handling done. This here should anyway only be a quick fix until we use
+		// the helm client lib. Then error handling will be better.
+		framework.HelmCmd(fmt.Sprintf("delete --purge %s", deploymentName))
+
+		f, err := ioutil.TempFile("/tmp", deploymentName)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		defer func() {
+			// Close & remove created tempfile.
+			fName := f.Name()
+			err = f.Close()
+			if err != nil {
+				// XXX: Tempfile leak.
```

Comment:
What is up with that?

Reply:
It's kind of a warning what would inevitably happen when above thing fails. If you think it's misleading, I can remove it. I just thought that it would be good to explicitly note what would happen if execution branches to `if`-body here.

Reply:
I may just not used to the brief expression as it is. I usually prefer more words.